lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -36,27 +34,45 @@ def execute(self):
             # We were not given any remotes on the command line, build the list
             # from the tools group directory.
             remotes = self.remote(self.tool_group_dir)
-
+            if not remotes:
+                # FIXME:  This is a weird case -- the group directory exists,
+                #  but contains no remotes. This is the result of clearing the
+                #  tools from a group which has already been cleared, because
+                #  the first clear does not remove the group directory.  This
+                #  behavior might be desirable for the `default` directory, but
+                #  it seems wrong for custom groups.
+                self.logger.error(f'No such group "{self.context.group}".')
+                return 1
+
+        tool_not_found = bool(self.context.name)  # Can't not find if not specified
```

Comment:
OK, I find your double negative vaguely disturbing; I also think this comment doesn't help much in understanding the logic. Perhaps something more like "if tool name is specified, assume it won't be found (until proven otherwise)"...

Reply:
I think it's actually a _triple_ negative.  (Negative boolean variables are always slightly tricky to describe succinctly.)

The problem is that I'm not sure that this one line is worth a block comment (if you think it is, I will happily provide one!), so it was a challenge to "say it all" in 30 or so characters, and that's what I came up with.  😆

I'm willing to remove the comment if you think it provides, um, negative value.  😉

Reply:
Actually, I find the logic of pre-supposing an error and canceling that error to be awkward. I'd much rather use a positive logic, although I understand the complication with trying to factor in the state of the option.

Still, your initial `bool(self.context.name)` isn't "tool not found" but "tool specified", whereas it *becomes* "tool found" at line 75. Which makes the final test really "if tool specified and not found"; but the implementation makes this relationship fuzzy and poorly documented.

It could be improved with just a name change, maybe, but I'd prefer to have the code more explicitly model the logic.

In any case, if you have trouble writing a short line comment describing something, it's probably time for (a) a block comment, or (b) cleaner logic that's easier to describe. 😆

Reply:
Speaking most generally, I think code should be self-documenting...so, if I need a **block** comment to explain it, it better be real important for it to be that way, or it should be re-written.  (I felt like that line would benefit from some esplainin', and then I was amused by the phrasing that the lack of space imposed on me, so I left it there to see if it would amuse you.)

But, clearly, this change is inappropriate, so I shall amend it.  🙂

(Although, I'm not coming up with a way to do it which uses "positive logic".  And, looking at the obvious name-change, I think I'll rework the logic instead.  😆)